### PR TITLE
libalx-base: Add new package

### PR DESCRIPTION
**Pull request by alejandro-colomar**
**libalx-base/1.0.0-b34**

Hi,

This is my personal library.  I would like to package it for conan to be able to use it like other conan packages.
This library is comprised of many modules, and I thought of packaging each one separately, like they do in Debian.  This package is for the base module, on which the rest depend.  However, if you think it should be packaged all in one single package, I can do that.

It's not obvious to me how to package it:
The build system is `make`, and it uses `pkg-config` to consume dependencies, and also to be consumed by users of the library.
But I've only seen examples here of libraries that use autotools or cmake.  I'd like to have some more basic example on which I could base this one, which also used only make and pkg-config.

So, I tried this first draft locally, but I got many errors, so I'll upload it and see if you can please help me develop this package.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.  `tried and failed, of course`.

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
>       ...
>       ERROR: 
>       	TypeError: 'NoneType' object is not callable
>       ```
> 

That's the same error I get locally.  Why is it?  I get it's a Python error, but I don't know much more.

**Comment by alejandro-colomar** [comment]
The build system in my library is set up so that you can analyze (static analysis), build, install and uninstall every module with separate commands.

The targets are `analyze`, `build`, `install`, `uninstall`, and `clean`, and you can use a suffix with the name of the module, to apply it to just a single module; for example `build-base`.

The make commands are run from `srcdir` (i.e., the root of the source repository).  The default build dir is `./tmp`, but it can be changed with `builddir=xxx`.  It supports [standard directory variables](https://www.gnu.org/software/make/manual/html_node/Directory-Variables.html).

For example, for the base package, to be analyzed and built in `/tmp/libalx-base/build/` and be installed into `/tmp/libalx-base/pkg/` with no `/usr/local` prefix for the `/lib` and `/include` dirs, and having downloaded the source code in `/tmp/libalx-base/src/` would be:

```
srcdir="/tmp/libalx-base/src";
builddir="/tmp/libalx-base/build";
DESTDIR="/tmp/libalx-base/pkg";
prefix="";
make -C ${srcdir} analyze-base builddir=${builddir};
make -C ${srcdir} build-base builddir=${builddir} DESTDIR=${DESTDIR} prefix=${prefix};
make -C ${srcdir} install-base builddir=${builddir} DESTDIR=${DESTDIR} prefix=${prefix};
```

The you just consume it by using `pkg-config --cflags --libs libalx-base`.

**Comment by alejandro-colomar** [comment]
It only supports Linux (if a BSD/Apple system provides GNU extensions, it may also work), but I didn't try.

There are no different build types (i.e. no release and debug); there's a single build.

The -fpic option is always on (both for static and shared libraries).
`build` generates both shared and static libraries.  You can then copy just install one of them or both.

I followed an approach similar to the Debian one.  I have `install-base-lib` for installing just the shared library; then an `install-base-dev` for installing header files (both C (`.h`) and C++ (`.hxx`)), the static library, the pkg-config (`.pc`) files, and some config files in some modules; and an `install-base-man` to install the manual pages.  I just realized that I didn't yet implement `install-base-doc`, which should install the LICENSE and maybe some other files.

**Comment by alejandro-colomar** [comment]
There are both C and C++ header files in all modules, and depending on the module, C source files, C++ source files, or a mix.

**Comment by alejandro-colomar** [comment]
I could further divide `install-base-dev` into `install-base-inc`, `install-base-a`,  `install-base-pc`, and `install-base-etc`, if it helps.

**Comment by alejandro-colomar** [comment]
BTW, I guess I'll have to remove the line installing the `.pc` file.  I'll keep it there for the moment, and when it complains that a `.pc` file is present I'll remove it.

**Comment by alejandro-colomar** [comment]
BTW, this library requires C2x (i.e., the future revision of the C standard) to compile.  I don't know if you support that.  In the Makefile it's already specified.

**Comment by alejandro-colomar** [comment]
Ping. @prince-chrismc, May you have a look at this, if you have some time?  :)

File: recipes/libalx-base/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "1.0.0-b34":
```
Comment (prince-chrismc) [inline recipes/libalx-base/all/conandata.yml:2]: We typically do not add prerelease

File: recipes/libalx-base/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from conans import ConanFile, tools
+
+
+class libalx_base_conan(ConanFile):
+    name = "libalx-base"
+    description = "C/C++ library providing useful extensions"
+    homepage = "https://github.com/alejandro-colomar/libalx"
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "LGPL-2.0-only"
+    topics = ("conan", "libc")
```
Comment (prince-chrismc) [inline recipes/libalx-base/all/conanfile.py:11]: I've seen conan removed recently

File: recipes/libalx-base/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        self._prefix   = ""
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version],
+                  destination=self._source_subfolder, strip_root=True)
+
+    def build(self):
+        srcdir   = self._srcdir
+        builddir = self._builddir
+        args = "builddir={}".format(builddir)
```
Comment (prince-chrismc) [inline recipes/libalx-base/all/conanfile.py:37]: Why the indirection? Could we not use the Conan variables directly

File: recipes/libalx-base/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import os
+from conans import ConanFile, tools
+
+
+class libalx_base_conan(ConanFile):
```
Comment (prince-chrismc) [inline recipes/libalx-base/all/conanfile.py:5]: Weird casing for python 

```suggestion
class LibalxConanFile(ConanFile):
```

File: recipes/libalx-base/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    name = "libalx-base"
+    description = "C/C++ library providing useful extensions"
+    homepage = "https://github.com/alejandro-colomar/libalx"
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "LGPL-2.0-only"
+    topics = ("conan", "libc")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {
+        "shared": [True, False],
```
Comment (prince-chrismc) [inline recipes/libalx-base/all/conanfile.py:14]: No fPIC?

**Comment by prince-chrismc** [comment]
> It also requires GCC 10 or later

It seems you are missing a validate step

**Comment by prince-chrismc** [comment]
I almost never work with C and Conan so I do not know of any specific settings... however I haven't seen any in CCI...

We have blocked compilers for missin parts of C99 so you can limit compilers here too

**Pull request closed without merging**